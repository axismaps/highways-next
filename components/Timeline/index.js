import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { range } from 'lodash';
import {
  Flex,
  Box,
  Heading,
  Text,
  IconButton,
  Slider,
  SliderTrack,
  SliderFilledTrack,
  SliderThumb,
} from '@chakra-ui/react';
import { ArrowLeftIcon, ArrowRightIcon } from '@chakra-ui/icons';

import config from '../../config';

const { minYear, maxYear } = config;
const roundedMinYear = Math.ceil(minYear / 10) * 10;
const roundedMaxYear = Math.floor(maxYear / 10) * 10;
const yearRange = range(roundedMinYear, roundedMaxYear, 10);

const Timeline = ({ handler }) => {
  const [year, setYear] = useState(1900);
  useEffect(() => {
    handler(year);
  }, [year]);
  return (
    <Flex
      py={[2, 5]}
      pr={5}
      pl={[5, 0]}
      boxShadow="0 2px 2px rgba(0,0,0,0.25)"
      pos="relative"
      zIndex={2}
      display={['block', 'flex']}
    >
      <Flex justifyContent="center">
        <IconButton
          icon={<ArrowLeftIcon />}
          mx={5}
          borderRadius="50%"
          onClick={() => setYear(Math.max(minYear, Math.round((year - 5) / 5) * 5))}
        />
        <Heading w="100px" textAlign="center">
          {year}
        </Heading>
        <IconButton
          icon={<ArrowRightIcon />}
          mx={5}
          borderRadius="50%"
          onClick={() => setYear(Math.min(maxYear, Math.round((year + 5) / 5) * 5))}
        />
      </Flex>
      <Slider
        colorScheme="gray"
        aria-label="slider-ex-1"
        value={year}
        min={minYear}
        max={maxYear}
        onChange={setYear}
        h="40px"
        mt={[2, 0]}
      >
        <Flex
          pos="relative"
          top="-12px"
          zIndex={1}
          w="100%"
          pl={`${((roundedMinYear - minYear) / (maxYear - minYear)) * 100}%`}
          pr={`${((maxYear - roundedMaxYear) / (maxYear - minYear)) * 100}%`}
        >
          {yearRange.map(y => (
            <React.Fragment key={y}>
              <Box
                borderLeft="1px solid black"
                boxSizing="border-box"
                w={`${100 / (yearRange.length - 1)}%`}
                h="27px"
                lineHeight="27px"
                mt="-2px"
                pl={1}
                userSelect="none"
                fontSize="0.7em"
              >
                <Text display={['none', 'block']}>{y}</Text>
              </Box>
            </React.Fragment>
          ))}
        </Flex>
        <SliderTrack h="30px">
          <SliderFilledTrack />
        </SliderTrack>
        <SliderThumb h="100%" w="16px" />
      </Slider>
    </Flex>
  );
};

Timeline.propTypes = {
  handler: PropTypes.func.isRequired,
};

export default Timeline;
