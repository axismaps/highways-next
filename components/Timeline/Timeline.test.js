import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import Timeline from './index';

describe('Timeline', () => {
  it('Renders the timeline', async () => {
    render(<Timeline handler={() => {}} />);
    expect(screen.getByText('1900')).toBeInTheDocument();
  });

  it('matches snapshot', async () => {
    const tree = renderer.create(<Timeline handler={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
