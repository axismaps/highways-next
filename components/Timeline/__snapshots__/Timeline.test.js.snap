// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline matches snapshot 1`] = `
.emotion-0 {
  display: block;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-right: 5px;
  box-shadow: 0 2px 2px rgba(0,0,0,0.25);
  position: relative;
  z-index: 2;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: none;
  width: auto;
  padding: 0px;
  border-radius: 50%;
  -webkit-margin-start: 5px;
  margin-inline-start: 5px;
  -webkit-margin-end: 5px;
  margin-inline-end: 5px;
}

.emotion-3 {
  width: 1em;
  height: 1em;
  display: inline-block;
  line-height: 1em;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: currentColor;
  vertical-align: middle;
}

.emotion-4 {
  width: 100px;
  text-align: center;
}

.emotion-7 {
  display: inline-block;
  position: relative;
  cursor: pointer;
  height: 40px;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  top: -12px;
  z-index: 1;
  width: 100%;
  padding-left: 1.8691588785046727%;
  padding-right: 0%;
}

.emotion-9 {
  border-left: 1px solid black;
  box-sizing: border-box;
  width: 5%;
  height: 27px;
  line-height: 27px;
  margin-top: -2px;
  padding-left: 1px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 0.7em;
}

.emotion-10 {
  display: none;
  display: block;
}

.emotion-51 {
  overflow: hidden;
  height: 30px;
}

.emotion-52 {
  width: inherit;
  height: inherit;
}

.emotion-53 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: absolute;
  outline: 0;
  height: 100%;
  width: 16px;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    <button
      className="chakra-button emotion-2"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="chakra-icon emotion-3"
        focusable={false}
        viewBox="0 0 24 24"
      >
        <g
          fill="currentColor"
        >
          <path
            d="M10.416,12a2.643,2.643,0,0,1,.775-1.875L20.732.584a1.768,1.768,0,0,1,2.5,2.5l-8.739,8.739a.25.25,0,0,0,0,.354l8.739,8.739a1.768,1.768,0,0,1-2.5,2.5l-9.541-9.541A2.643,2.643,0,0,1,10.416,12Z"
          />
          <path
            d="M.25,12a2.643,2.643,0,0,1,.775-1.875L10.566.584a1.768,1.768,0,0,1,2.5,2.5L4.327,11.823a.25.25,0,0,0,0,.354l8.739,8.739a1.768,1.768,0,0,1-2.5,2.5L1.025,13.875A2.643,2.643,0,0,1,.25,12Z"
          />
        </g>
      </svg>
    </button>
    <h2
      className="chakra-heading emotion-4"
    >
      1950
    </h2>
    <button
      className="chakra-button emotion-2"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="chakra-icon emotion-3"
        focusable={false}
        viewBox="0 0 24 24"
      >
        <g
          fill="currentColor"
        >
          <path
            d="M13.584,12a2.643,2.643,0,0,1-.775,1.875L3.268,23.416a1.768,1.768,0,0,1-2.5-2.5l8.739-8.739a.25.25,0,0,0,0-.354L.768,3.084a1.768,1.768,0,0,1,2.5-2.5l9.541,9.541A2.643,2.643,0,0,1,13.584,12Z"
          />
          <path
            d="M23.75,12a2.643,2.643,0,0,1-.775,1.875l-9.541,9.541a1.768,1.768,0,0,1-2.5-2.5l8.739-8.739a.25.25,0,0,0,0-.354L10.934,3.084a1.768,1.768,0,0,1,2.5-2.5l9.541,9.541A2.643,2.643,0,0,1,23.75,12Z"
          />
        </g>
      </svg>
    </button>
  </div>
  <div
    className="chakra-slider emotion-7"
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0,0,0,0)",
        "outline": 0,
        "paddingBottom": 0,
        "paddingTop": 0,
        "position": "relative",
        "touchAction": "none",
        "userSelect": "none",
      }
    }
    tabIndex={-1}
  >
    <div
      className="emotion-8"
    >
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1840
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1850
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1860
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1870
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1880
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1890
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1900
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1910
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1920
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1930
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1940
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1950
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1960
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1970
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1980
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          1990
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          2000
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          2010
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          2020
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          2030
        </p>
      </div>
      <div
        className="emotion-9"
      >
        <p
          className="chakra-text emotion-10"
        >
          2040
        </p>
      </div>
    </div>
    <div
      className="chakra-slider__track emotion-51"
      id="slider-track-3"
      style={
        Object {
          "position": "absolute",
          "top": "50%",
          "transform": "translateY(-50%)",
          "width": "100%",
        }
      }
    >
      <div
        className="chakra-slider__filled-track emotion-52"
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "top": "50%",
            "transform": "translateY(-50%)",
            "width": "53.271028037383175%",
          }
        }
      />
    </div>
    <div
      aria-label="slider-ex-1"
      aria-orientation="horizontal"
      aria-valuemax={2050}
      aria-valuemin={1836}
      aria-valuenow={1950}
      className="chakra-slider__thumb emotion-53"
      id="slider-thumb-3"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="slider"
      style={
        Object {
          "left": "calc(53.271028037383175% - 0px)",
          "position": "absolute",
          "touchAction": "none",
          "userSelect": "none",
        }
      }
      tabIndex={0}
    />
    <input
      type="hidden"
      value={1950}
    />
  </div>
</div>
`;
